adding turkey gitops

tanfarming opened the pull request:
- added Dockerfile that builds nginx hosted static files packaged as per reticulum needs
  - container accepts following parameters
    - DOMAIN
    - SUB_DOMAIN
    - turkeyCfg_base_assets_path
    - turkeyCfg_non_cors_proxy_domains
    - turkeyCfg_reticulum_server
    - turkeyCfg_cors_proxy_server

robinkwilson commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+run apk add bash
+run mkdir /ssl && mkdir -p /www/hubs && mkdir -p /www/hubs/pages && mkdir -p /www/hubs/assets
+copy --from=ssl /ssl /ssl
+copy --from=builder /hubs/rawhubs/pages /www/hubs/pages
```
I'd add rawhubs/ to the `.gitignore` file.

robinkwilson commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+/healthcheck.sh& \n\
+nginx -g "daemon off;" \n\n' > /run.sh
+run chmod +x /run.sh && cat /run.sh
+cmd bash /run.sh
```
Where is the run.sh file?

netpro2k replied:
It is generated in the lines above. Though I suggested we should actually pull this out and not generate it at build time.

robinkwilson commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
```
nit, if it's relevant: Are there any dependencies or instructions to add to run this in the README?

robinkwilson reviewed (commented):
beginner docker question: is there a file depicting all of the dependencies like node or npm versions that hubs runs on? Is that included here or another place.

robinkwilson commented:
Ah I see this is only as it relates to turkey. So this is NOT relevant to our users wanting to run hubs using a docker file. 

Can you add a comment at the top of this docker file explaining it.

netpro2k commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+run mkdir rawhubs && mv dist/pages rawhubs && mv dist/assets rawhubs && mv dist/react-components rawhubs/pages && mv dist/favicon.ico rawhubs/pages
+
+from alpine/openssl as ssl
+run mkdir /ssl && openssl req -x509 -newkey rsa:2048 -sha256 -days 36500 -nodes -keyout /ssl/key -out /ssl/cert -subj '/CN=hubs'
```
What is this for? I guess this is for ssl between reticulum and the hubs source assets, though these are self signed certs so I guess it has to ignore their validity? Is this just to ensure encryption over the wire? These are public assets anyway I think, so this nginx service can probably just be running http. I might also just be missing something obvious.

tanfarming replied:
long story short / if i remembered correctly -- it's for ret, it only wants assets served on https
we technically can terminate https on ingress (fraefik / haproxy) otherwise, i just didn't want to bother with more ret code changes lol

brianpeiris replied:
I think changing ret to remove the ssl requirement is worth it later on, but it's okay to do this for now.

netpro2k reviewed (commented):
This is looking good. I would pull out the two .sh files rather than generating them in the dockerfile and I had some questions re-caching to make sure we make builds as fast as possible. Getting `npm ci` in particular to only run when actually necessary would drastically cut down build times. Also was a bit confused why the ssl stuff is needed.

tanfarming commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
+from node:lts as builder
+workdir hubs
+copy . .
```
yeah i was thinking to run in only in github action's hosted agent so caching's not applicable there, we do have option to use github action's caching feature but i'm not yet sure if the few minutes saved would worth the cost of extra pipeline complexity + losing the portability of wrapping everything inside dockerfile
I think for local testing a custom Dockerfile should be used, and depends on the individual developer it could be different, maybe we can later add an example in the readme for it?

netpro2k replied:
Ah yeah, I assumed github actions just had built in handling for this... 

Some quick searching shows it shouldn't be too hard to setup https://github.com/progapandist/anycable_rails_demo/blob/master/.github/workflows/fake_deploy_multistage.yml (from https://evilmartians.com/chronicles/build-images-on-github-actions-with-docker-layer-caching) by using github's cache action to cache the build cache directory. 

Re  "extra pipeline complexity + losing the portability of wrapping everything inside dockerfile" I don't think this moves anything out of the dockerfile, just configures the environment it is running in. Not the highest priority ever, but shaving off build time definitely adds up. Our current CI builds take quite a long time, and its highly dominated by `npm ci` time which actually changes quite rarely. Even if we don't go through setting up github actions to enable caching, making this part of the dockerfile more cache friendly seems good (unless there is some tradeoff)

tanfarming replied:
yup agreed i'll add the github action caching -- just discovered yesterday that file ops are like 4-6 times slower in "docker file system" so the original ~6 min build time becomes ~25 min inside docker
the extra github action caching step totally worth it in this case

tanfarming commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+copy --from=builder /hubs/rawhubs/pages /www/hubs/pages
+copy --from=builder /hubs/rawhubs/assets /www/hubs/assets
+run echo "server {listen 8080 ssl;ssl_certificate /ssl/cert;ssl_certificate_key /ssl/key; location / {root /www;autoindex off;add_header 'Access-Control-Allow-Origin' '*';}}" > /etc/nginx/conf.d/default.conf
+run printf 'while true; do (echo -e "HTTP/1.1 200 OK\r\n") | nc -lp 1111 > /dev/null; done' > /healthcheck.sh && chmod +x /healthcheck.sh
```
yep will do

tanfarming commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+from alpine/openssl as ssl
+run mkdir /ssl && openssl req -x509 -newkey rsa:2048 -sha256 -days 36500 -nodes -keyout /ssl/key -out /ssl/cert -subj '/CN=hubs'
+
+from nginx:alpine
```
jargon's "multi stage docker build"
mainly to slim down the final product, ie we need sdk to build, but just runtime to host
i haven't optimized the dockerfile yet, ie security, further slim-downs, etc, this push is just to get the devops started

netpro2k replied:
Cool. Will read up on that.

netpro2k commented:
@robinkwilson 
> beginner docker question: is there a file depicting all of the dependencies like node or npm versions that hubs runs on? Is that included here or another place.

The `from node:16.13 as builder` line in the Dockerfile is responsible for what version of node/npm gets used during the build. At runtime its just static assets so only nginx is involved.

tanfarming commented:
> @robinkwilson
> 
> > beginner docker question: is there a file depicting all of the dependencies like node or npm versions that hubs runs on? Is that included here or another place.
> 
> The `from node:16.13 as builder` line in the Dockerfile is responsible for what version of node/npm gets used during the build. At runtime its just static assets so only nginx is involved.
yep, not only node version, OS and all other files/packages are pinned down too -- it's basically a "docker managed VM", cloud jargon here is "immutable infra"
with npm ci / package-lock.json there's no moving parts other than our code

tanfarming commented:
@netpro2k @robinkwilson

updated:
- changed dockerfile's name and added comments
- inline run.sh moved to ./scripts/docker/
- inline healthcheck.sh become a function inside run.sh
- docker cache layers on github action enabled with buildx + type=registry after wasted days on the buggy docker/buildx/buildKit
  - type=local + diy move files: hackish, extra steps, a bit slower
  - type=gha: docker/buildx wants us to use this, still "experimental"
  - type=registry: seems to be the cleanest
  - docker's buildx' still buggy, [for example](https://github.com/mozilla/hubs/runs/4193826833?check_suite_focus=true#step:5:178), hope it will get better
  - docker file system and "non-ancient" npm versions are not playing well together, it seems npm runs many times slower inside docker, but lets worry about this later
    - [npm on agent with fresh npm ci: ~8min](https://github.com/mozilla/turkey-ops/actions/runs/1199621038)
    - [npm inside docker file system with fresh npm ci: ~33min](https://github.com/mozilla/hubs/actions/runs/1453911447)
    - [npm inside docker file system with npm ci cached: ~10min](https://github.com/mozilla/hubs/actions/runs/1454494526)

johnshaughnessy commented on RetPageOriginDockerfile:
```diff
@@ -0,0 +1,30 @@
+run mkdir -p /hubs/admin/ && cd /hubs
+copy package*.json ./
+copy admin/package*.json admin/
+run npm ci
```
Should the `package-lock.json` checked into the repo be built with node 16.13 to prevent the following?
```
Step 5/24 : run npm ci
 ---> Running in 45dcecf84c37
npm WARN old lockfile
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile
npm WARN old lockfile This is a one-time fix-up, please be patient...
```

tanfarming replied:
yep, but we'll do it in future when we are updating hubs' dependencies, this is just an initial devops update for turkey and i'd like to minimize any disturbance to the existing codebase

brianpeiris replied:
I think this might be one of the reasons the build is slow. That "one-time fix-up" will happen every time and it downloads quite a lot of metadata, which slows things down. As you noted earlier, it is much faster with an older version of node/npm.

But I agree, this is a problem to be solved on the hubs side, not a problem with this workflow, so it can be tackled at a later time.

johnshaughnessy commented on RetPageOriginDockerfile:
```diff
@@ -0,0 +1,30 @@
+from node:16.13 as builder
+run mkdir -p /hubs/admin/ && cd /hubs
+copy package*.json ./
+copy admin/package*.json admin/
```
nit: Since we only seek to copy `package.json` and `package-lock.json`, maybe these should list them explicitly instead of catching them with a wildcard:
```
copy package.json ./
copy package-lock.json ./
copy admin/package.json admin/
copy admin/package-lock.json admin/
```

tanfarming replied:
yep will do

johnshaughnessy commented:
> docker's buildx' still buggy, [for example](https://github.com/mozilla/hubs/runs/4193826833?check_suite_focus=true#step:5:178), hope it will get better

What bug you are showing in that example? I see the same output spammed repeatedly
```
#22 [builder  6/11] RUN cd admin && npm ci && cd ..
#22 CACHED
```
but don't know enough about buildx to know how to interpret this.

tanfarming commented:
> > docker's buildx' still buggy, [for example](https://github.com/mozilla/hubs/runs/4193826833?check_suite_focus=true#step:5:178), hope it will get better
> 
> What bug you are showing in that example? I see the same output spammed repeatedly
> 
> ```
> #22 [builder  6/11] RUN cd admin && npm ci && cd ..
> #22 CACHED
> ```
> 
> but don't know enough about buildx to know how to interpret this.
yeaht that's a bad example, i saw the step 6/11 repeated so many times and got confused and thought it's a bug, googled it and find some other guys online also got confused
until yesterday Brian and Dom pointed out to me that it's probably just downloading
I'm pretty green with their buildkit/buildx too, tried it out a few years ago, but went back to vanilla docker build, 
there are other issues, I ran into a few of them last weekend, for example this one https://github.com/moby/buildkit/issues/1939
we'll see how it goes, I'm using buildkit because there's fewer steps and also it seems to be where things are going (with future native github action support etc) -- we do have other options, like using vanilla docker build with diy file copying etc to back up/restore local caches with github action's "file caches"

brianpeiris commented on scripts/docker/run.sh:
```diff
@@ -0,0 +1,19 @@
+for f in /www/hubs/pages/*.html; do 
+    for var in $(printenv); do 
+    var=$(echo $var | cut -d"=" -f1 ); prefix="turkeyCfg_"; 
+    [[ $var == $prefix* ]] && sed -i "s/$anchor/ <meta name=\"env:${var#$prefix}\" content=\"${!var//\//\\\/}\"\/> $anchor/" $f; 
```
```sh
${!var//\//\\\/}
```
Is this substitution actually doing what you intended? I stared at it for a while and tried to run it myself in a terminal, but I can't quite understand what the intent was. In fact, I think it might be wrong at the moment. Might be helpful to explain it in a comment in this script as well, since it's hard to read without thinking about it for a while.

tanfarming replied:
yeah substitution part is just swapping `/` for `\/`, for the slashes in `base_assets_path`
both sed and bash parameter substitution do not like `/` so it needs to be escaped into `\/`
sed does not like `\` either, so the `\/` part end up got escaped into `\\\/`
